src/System.Data.Extensions/src/System/Data/DataRowComparer.cs

```diff
@@ -0,0 +1,238 @@
+                return true;
+            }
+
+            if (ReferenceEquals(a, null) ||
+                ReferenceEquals(b, null))
+            {
+                return false;
+            }
+
+            if (a.Length != b.Length)
+            {
+                return false;
+            }
+
+            for (int i = 0; i < a.Length; ++i)
+            {
+                if (!a[i].Equals(b[i]))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    /// <summary>
+    /// This class implements IEqualityComparer using value based semantics
+    /// when comparing DataRows.
+    /// </summary>
+    public sealed class DataRowComparer<TRow> : IEqualityComparer<TRow> where TRow : DataRow
+    {
+        /// <summary>
+        /// Private constructor to prevent initialization outside of Default singleton instance.
+        /// </summary>
+        private DataRowComparer() { }
+
+        private static DataRowComparer<TRow> s_instance = new DataRowComparer<TRow>();
+
+        /// <summary>
+        /// Gets the singleton instance of the data row comparer.
+        /// </summary>
+        public static DataRowComparer<TRow> Default { get { return s_instance; } }
+
+        /// <summary>
+        /// This method compares to DataRows by doing a column by column value based
+        /// comparision.
+        /// </summary>
+        /// <param name="leftRow">The first input DataRow</param>
+        /// <param name="rightRow">The second input DataRow</param>
+        /// <returns>True if rows are equal, false if not.</returns>
+        public bool Equals(TRow leftRow, TRow rightRow)
+        {
+            if (ReferenceEquals(leftRow, rightRow))
+            {
+                return true;
+            }
+
+            if (ReferenceEquals(leftRow, null) ||
+                ReferenceEquals(rightRow, null))
+            {
+                return false;
+            }
+
+            if (leftRow.RowState == DataRowState.Deleted || rightRow.RowState == DataRowState.Deleted)
+            {
+                throw DataSetUtil.InvalidOperation(SR.DataSetLinq_CannotCompareDeletedRow);
+            }
+
+            int count = leftRow.Table.Columns.Count;
+            if (count != rightRow.Table.Columns.Count)
+            {
+                return false;
+            }
+
+            for (int i = 0; i < count; ++i)
+            {
+                if (!DataRowComparer.AreEqual(leftRow[i], rightRow[i]))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// This method retrieves a hash code for the source row.
+        /// </summary>
+        /// <param name="row">The source DataRow</param>
+        /// <returns>HashCode for row based on values in the row.</returns>
+        public int GetHashCode(TRow row)
+        {
+            DataSetUtil.CheckArgumentNull(row, "row");
```

Comment:
Nit: We might want to leave cleanup for a subsequent PR or set of PRs, but this could be changed from `"row"` to `nameof(row)`.